### DISCUSSION TITLE
Required is great but needs more!

### DISCUSSION THREAD

**Discussion by JohnGalt1717** [dotnet/csharplang#6678] [Language Ideas]
I am constantly doing something like this with records (same issue in a variant occurs with setable properties on a class or record too):

```
public Document CreateDocFromDto(DocumentDto dto) {
  var doc = await dbSession.xyz.FirstOrDefault(d => d.Id == SomeId);
  doc ??= new Document() {
    {required fields that are likely in the mutations under the next section}
  }
  
  return document with {
     {Mutations from say a dto or something}
  }
}

```
Required fixes the problems of null! or other bad stuff.  But it induces you to repeat yourself, but if you have required fields, what happens is that in this function, if you new up the document you have to pass the required fields AND then if it wasn't a new and it was in the database you still don't have those values so you have to do the document with and have the same fields which is repetition.

I'd like to see this work:

```
public Document CreateDocFromDto(DocumentDto dto) {
  var doc = await dbSession.xyz.FirstOrDefault(d => d.Id == SomeId);
  doc ??= new Document();
  
  return document ?? new Document() with {
     {Mutations from say a dto or something}
  }
}
```

The new Document() line should only error if the required fields aren't in the with, otherwise it should accept it because you're defining all of the required in the with statement.

In the background this would do:

```
if (document == null) {
   return new Document() { {all of the assignments that are in the with}};
} else {
   return document with {all of the assignments in the with}};
}
```

Obviously this would be even nicer:

```
return ?== new Document() with {
   {with assignments}
}

```
?== would mean, if null then assign to new and either way, apply the with. And of course if there were required that weren't in the with, then you could put them in the assignment of the new Document() section in either case of the suggestion.

Here's an actual example with optimization checking for the Id which isn't really needed for the example, but...

```
	public static async Task<EventListener> FromDto(IPltfrmdSession dbSession, EventListenerDto dto, CancellationToken cancellationToken = default) {
		var document = dto.Id is null ? null : await dbSession.EventListeners.GetAsync(dto.Id, PltfrmdUser.Empty, true, cancellationToken);

		document ??= new EventListener(dto.Id) {
			SourceType = dto.SourceType
		};

		return document with {
			SourceType = dto.SourceType,
			DestinationType = dto.DestinationType,
			EventTypes = dto.EventTypes.ToImmutableList()
		};
	}
```

This is all over my code, and is just repetitious.

**Comment by JohnGalt1717** [comment]
Along with this it would be nice to do:

public required readonly record struct SomeDto {}

Meaning that without having to create a constructor, all fields must be specified when being created unless otherwise given a value or nullable.

**Comment by 333fred** [comment]
I'm not sure that I understand the scenario here. In particular, given this example:

```cs
public Document CreateDocFromDto(DocumentDto dto) {
  var doc = await dbSession.xyz.FirstOrDefault(d => d.Id == SomeId);
  doc ??= new Document(); // I'm assuming this isn't necessary, given the ?? below?
  
  return document ?? new Document() with {
     {Mutations from say a dto or something}
  }
}
```

The new `Document` DTO is going to have all of the required members missing. What will fill them in and ensure this new DTO is in a valid state? When I do `CreateDocFromDto(myDto).Text.Substring(1);`, what is going to ensure that `Text` isn't null?

**Reply by JohnGalt1717** [reply]
The issue is that return document ?? new Document() with { Other fields} will throw an error.

So you'd have to write it as:

document ?? new Document { {Required fields} } with { {mutations that repeat required fields} } which is redundant and violates DRY.

I should only have to put in {Required Fields} that aren't in the with  that is immediately after it.

**Reply by 333fred** [reply]
I think I understand where we were talking past each other: you mean `(document ?? new Document()) with { ... }`, not `document ?? (new Document() with { ... })`. The latter version is how the compiler will interpret `document ?? new Document() with { ... }`.

In general, I'm a bit skeptical of this use case. What guarantees that the set of mutations in the DTO is actually _all_ the required members? What if only `Document.FileName` was modified, and `Document.Text` was unmodified? My approach to this problem is to have a general method for constructing a `Document` from a `DocumentDTO`, and then just return a call to that method when `FindAsync` returns null. And then for the case where it does return something, do a `with` as appropriate. Yes, it does mean that the assignment code is duplicated for the mutations, but I don't see a way to make it work with partial assignment otherwise.

In particular for `with`, this approach just won't work correctly. `with` starts by creating a _new_ object, so that `new Document()` is just thrown away immediately, all of its (empty) fields are copied, and then the new assignments in the `with` are executed. And that copy process is actually user-modifiable, as you can define your own copy constructor. That ctor might do something that relies on the original being a valid instance of `Document`, and this would break that.

**Comment by svick** [comment]
I think what you want isn't:

```c#
return document ?? new Document() with {
   {Mutations from say a dto or something}
}
```

But:

```c#
return (document ?? new Document()) with {
   {Mutations from say a dto or something}
}
```

I.e. apply the `with` to the whole `document ?? new Document()`, not just to `new Document()`.

Though this still doesn't compile, because the compiler does not recognize `with` as part of the construction of `new Document()`.

**Reply by JohnGalt1717** [reply]
Both are the same thing because ?? applies before with in order of operations in C#...

And yes, this is exactly the case that should work and is very common.

Given that the other side is either a return or an assignment back to the same variable 99% of the time when using "with" ?== makes it very elegant.

**Reply by 333fred** [reply]
> Both are the same thing because ?? applies before with in order of operations in C#...

No it doesn't. Sorry :).

**Reply by 333fred** [reply]
As a simple example: https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYAMBYAUBgRjzwDcBDMAAmAFMBnYSgXkqhoHdKAVe4ACgCUlAN6UAys0oAiAlMoBfANx5UBAJx9aDSgH4drDt16DK7AJbAAFiPGSp6OfIHL86vpt679bTjwaChcysbCRZ7R2difABmSjAaAGMAezA4IwY8YTxKHMpUWNVMW1EAcxpgRUozKAtK+Tx5IA===

**Reply by JohnGalt1717** [reply]
If that is indeed the case (it doesn't seem to be in my testing) then even more reason for ?==.
